models/ctsm/R/write_params-clm.R

```diff
@@ -1,670 +1,103 @@
-      # low_moisture_S	Slope of fuel moisture to burned fraction term for drier fuel	litterclass	
-      # max_decomp	Maximum decomposition rate of litter in the absence of moisture or temperature stress, per fuel class	litterclass	y-1
-      # mid_moisture	Parameter of burned fraction term. Below this 'low' constants apply, above this, 'mid' constants apply, 	litterclass	
-      # mid_moisture_C	Intercept (constant) of fuel moisture to burned fraction term for wetter fuel	litterclass	
-      # min_moisture	Parameter of burned fraction term. Below this value all litter is burned by a fire. Above, 'low' constants apply	litterclass	
-      # FBD	Fuel Bulk Density of fuel class 	litterclass	kg m-3
-      # alpha_FMC	Parameter of function relating fuel moisture content to meteorological fire danger index 	litterclass	
-      # SAV	Surface Area to Volume Ratio of fuel class 	litterclass	cm-1
-      
-      ## NCWD dimensioned       Size:4
-      if(var == "CWD_frac1"){            ##Fraction of coarse woody debris (CWD) that is moved into each of the four woody fuel classes
-        ncdf4::ncvar_put(nc=fates.param.nc, varid='fates_CWD_frac', start = 1, count = 1,
-                         vals=pft[v])
-      }
-      if(var == "CWD_frac2"){
-        ncdf4::ncvar_put(nc=fates.param.nc, varid='fates_CWD_frac', start = 2, count = 1,
-                         vals=pft[v])
-      }
-      if(var == "CWD_frac3"){
-        ncdf4::ncvar_put(nc=fates.param.nc, varid='fates_CWD_frac', start = 3, count = 1,
-                         vals=pft[v])
-      }
-      if(var == "CWD_frac4"){
-        ncdf4::ncvar_put(nc=fates.param.nc, varid='fates_CWD_frac', start = 4, count = 1,
-                         vals=pft[v])
-      }
-      
-      
-    } ## end loop over VARIABLES
-  } ## end loop over PFTs
-  #ncdf4::nc_close(param.nc)
-  ncdf4::nc_close(clm.param.nc)
+#' /dontrun{
+#' trait.values <- list(c4_grass = list(sla = 0.03, c2n_leaf = 35, stom_slope = 10, fineroot2leaf = 2, vcmax = 50))
+#' write_params_ctsm(trait.values = trait.values, run.id = 1)
+#' }
 
-  
-}
+# TODO: update this to read param files (CTSM and FATES) out of the refcase directory, not the PEcAn package
+# TODO: update to allow it to pick between CLM4.5 and CLM5 (CLM5 == CTSM) parameter set based on refcase, user selection
+# TODO: match Pecan pft names with CTSM pft names when Pecan name is not identical to a CTSM name
+## See issue https://github.com/PecanProject/pecan/issues/1008
 
+write_params_ctsm <-
+  function(defaults = system.file('clm5_params.c171117_0001.nc', package = 'PEcAn.CTSM'),
+           trait.values,
+           settings,
+           run.id) {
+    
+    ## Copy and open default parameter files
+    ctsm.param.default <-
+      system.file('clm5_params.c171117_0001.nc', package = 'PEcAn.CTSM')
+    if (!exists('local.rundir'))
+      local.rundir <- tempdir()
+    if (!exists('run.id'))
+      run.id <- 1
```

Comment:
`local.rundir` should really be an argument to this function, with a default value of `NULL`. More generally, our preferred approach for defaults like this to either assign the default directly in the signature, or (for more complex default behavior) to set the default to `NULL` and test for that:

```r
write_params_ctsm <- function(..., run.id = 1, local.rundir = tempdir()) {...}
# ...or...
write_params_ctsm <- function(..., run.id = NULL, local.rundir = NULL) {
  if (is.null(run.id)) run.id <- 1
  if (is.null(local.rundir)) local.rundir <- tempdir()
  # ...
}
```

Reply:
Went with the simpler option for now!